doc/sphinx/source/libCEEDdev.md

```diff
@@ -1,98 +1,109 @@
-```
+The `/gpu/*/shared` backends delegate to the corresponding `/gpu/*/ref` backends.
+These backends use shared memory to reduce the amount of memory required for the {ref}`CeedBasis` kernels.
 
-means that it can be traversed as
+The `/gpu/*/gen` backends delegate to the corresponding `/gpu/*/shared` backends.
+These backends write a single comprehensive kernel to apply the action of the {ref}`CeedOperator`, significantly improving performance by eliminating intermediate data structures and reducing the total number of kernel launches required.
 
-```c
-for (d=0; d<dim; d++)
-  for (c=0; c<num_comp; c++)
-    for (q=0; q<Q; q++)
-      for (e=0; e<num_elem; e++)
-        u[((d*num_comp + c)*Q + q)*num_elem + e] = ...
-```
+The `/gpu/*/magma` backends delegate to the corresponding `/gpu/*/ref` backends.
+These backends provide better performance for {ref}`CeedBasis` kernels but do not have the improvements from the `/gpu/*/gen` backends for {ref}`CeedOperator`.
 
-This ordering is sometimes referred to as row-major or C-style.
-Note that flattening such as
+The `/*/*/occa` backends are not part of any family and generally are not well maintained.
 
-```c
-// u has shape [dim, num_comp, Q*num_elem]
-```
+## Internal Layouts
 
-and
+Ceed backends are free to use any **E-vector** and **Q-vector** data layout, to include never fully forming these vectors, so long as the backend passes the `t5**` series tests and all examples.
+There are several common layouts for **L-vectors**, **E-vectors**, and **Q-vectors**, detailed below:
 
-```c
-// u has shape [dim*num_comp, Q, num_elem]
-```
+- **L-vector** layouts
 
-are purely implicit -- one just indexes the same array using the appropriate convention.
+  - **L-vectors** described by a {ref}`CeedElemRestriction` have a layout described by the `offsets` array and `comp_stride` parameter.
+    Data for node `i`, component `j`, element `k` can be found in the **L-vector** at index `offsets[i + k*elem_size] + j*comp_stride`.
+  - **L-vectors** described by a strided {ref}`CeedElemRestriction` have a layout described by the `strides` array.
+    Data for node `i`, component `j`, element `k` can be found in the **L-vector** at index `i*strides[0] + j*strides[1] + k*strides[2]`.
 
-## `restrict` Semantics
+- **E-vector** layouts
 
-QFunction arguments can be assumed to have `restrict` semantics.
-That is, each input and output array must reside in distinct memory without overlap.
+  - If possible, backends should use {c:func}`CeedElemRestrictionSetELayout()` to use the `t2**` tests.
+    If the backend uses a strided **E-vector** layout, then the data for node `i`, component `j`, element `k` in the **E-vector** is given by `i*layout[0] + j*layout[1] + k*layout[2]`.
+  - Backends may choose to use a non-strided **E-vector** layout; however, the `t2**` tests will not function correctly in this case and the tests will need to be whitelisted for the backend to pass the test suite.
+
+- **Q-vector** layouts
```

Comment:
The documentation for the Q-vector layout and the L-vector layout should also be in "normal" documentation somewhere, since those are directly user facing. I know the L-vector layouts are described in the "normal" and "strided" restriction functions, but I don't think Q-vector is described anywhere.

Reply:
looks like the shapes are described in `CeedQFunctionUser`